### DISCUSSION TITLE
New controlling plugin: Amount of working hours

### DISCUSSION THREAD

**Discussion by idemus** [kimai/Paid-plugins#112] [Questions & Answers] [answered]
Hi,
we have different customers and for each customer we have different projects, activities to create billable timesheets.
The working contract has it's own project for the internal customer (our own company).
The working time overview under the menue working contract sums up all activities from all customers/projects. 
Is it possible to set a default 'Customer/Project' in the user properties "working contract" which is then taken into account for the working hours overview?

Best regards
Klaus

**Comment by kevinpapst** [comment]
> The working contract has it's own project for the internal customer (our own company).

I don't know what this means. Are you talking about the old MR plugin?
This is not the old plugin 😁  the new one works different: all booked time is working time.

~~If you want to track attendance and project times in parallel, you need to use two distinct user accounts for now.~~

EDIT: There is now a setting to use one specific project globally for working time.

**Comment by idemus** [comment]
That was the same way in the old MR plugin. 
A second user account is not a solution. Would it be possible  that you'll implement it in future as a feature?

**Reply by kevinpapst** [reply]
> A second user account is not a solution. 

It is the only solution if you want to mix attendance and project related times in parallel in Kimai.

> Would it be possible that you'll implement it in future as a feature?

I am sorry that the MR plugin got you hooked on this broken idea, which has way too many downsides.
The definitive answer is: No.

**Reply by idemus** [reply]
I'll try on my own with a "modification" in your coding. I know that I'll loose it when there are updates without manually updating my insertion again and again...
Could you please tell me where I can find the query which fetches the entries for the working time overview?

**Reply by kevinpapst** [reply]
Start searching from here: https://github.com/kimai/kimai/blob/main/src/WorkingTime/WorkingTimeService.php#L51

**Reply by kevinpapst** [reply]
And please share your changes, who knows ...

**Reply by idemus** [reply, marked as answer]
Hi Kevin,

here my changes I've done in WorkingTimeService.php:
- adding third parameter $project in function     private function getYearStatistics(\DateTimeInterface $year, User $user, int $project ): array
- adding addwhere in the select:
         ->andWhere($qb->expr()->eq('t.project', ':project'))
         ->setParameter('project', $project)

- calling the function in line 71 with projectid (hard coded because it's only for  my own use).
$stats = $this->getYearStatistics($yearDate, $user, 4);

Now I get only those timesheets created under my project which is considered for working time (and not times booked on customer projects for billable reasons).

Best regards
Klaus

**Reply by kevinpapst** [reply]
Really that easy? In that case, I might add it to core. 
Please let me know if you find any issues, so I can help improving it

**Reply by Wbrugger** [reply]
Hello,

this is exactly my Problem with the new Plugin. Please adopt the solution from idemus quick. I have a own feature request #175.

Best Regards,

WBrugger

**Comment by Wbrugger** [comment]
Special thx to idemus :)

I have changed the code in the Module like him. But i filtered by activity not by project.

71              if ($stats === null) {
                    $stats = $this->getYearStatistics($yearDate, $user, 6); // Added activity Hardcoded
                }

143           $qb
144         ->select('COALESCE(SUM(t.duration), 0) as duration')
145         ->addSelect('DATE(t.date) as day')
146          ->where($qb->expr()->isNotNull('t.end'))
147          ->andWhere($qb->expr()->between('t.date', ':begin', ':end'))
148           ->andWhere($qb->expr()->eq('t.user', ':user'))
149 		->andWhere($qb->expr()->eq('t.activity', ':activity'))  //try to filter activity
150            ->setParameter('end', $end->format('Y-m-d'))
152            ->setParameter('user', $user->getId())
153		->setParameter('activity', $activity)  //Set activity Value
154           ->addGroupBy('day')
155        ;

I have tested the chnges. Working like a charme :)

Best Regards,

WBrugger

**Comment by wobouser** [comment]
thanks for the input to @idemus and @Wbrugger
that would have been a good interim solution for us.

unfortunately, neither solution works for me.
I don't know if something has been fundamentally changed in the WorkingTimeService.php so far (an indication of this is that the line numbers from the examples no longer match the current code).
I have tried both ways (actvities & projects).

my wish would actually be that there is the possibility to record the working time as a project or activity.
we also have times that cannot be productively assigned to a customer.
so it is important for us to have a come and go time in order to record attendance in the company due to legal regulations

**Reply by idemus** [reply]
@wobouser 
Here is my coding I've inserted in file WorkingTimeService.php (search for "insertion") which is working with plugin version 1.8.
Maybe this could help you:
[...]
    public function getYear(User $user, \DateTimeInterface $yearDate, \DateTimeInterface $until): Year
    {
        $yearTimes = $this->workingTimeRepository->findForYear($user, $yearDate);
        $existing = [];
        foreach ($yearTimes as $workingTime) {
            $existing[$workingTime->getDate()->format('Y-m-d')] = $workingTime;
        }

        $year = new Year(\DateTimeImmutable::createFromInterface($yearDate), $user);

        $stats = null;
        $firstDay = $user->getWorkStartingDay();

        foreach ($year->getMonths() as $month) {
            foreach ($month->getDays() as $day) {
                $key = $day->getDay()->format('Y-m-d');
                if (\array_key_exists($key, $existing)) {
                    $day->setWorkingTime($existing[$key]);
                    continue;
                }

                if ($stats === null) {
                **/* begin of insertion
                   filter working time overview to a specific project */
                    $stats = $this->getYearStatistics($yearDate, $user, 4);
                /* end of insertion */**
                }

                $dayDate = $day->getDay();
                $result = new WorkingTime($user, $dayDate);

                if ($firstDay === null || $firstDay <= $dayDate) {
                    $result->setExpectedTime($user->getWorkHoursForDay($dayDate));
                }

                if (\array_key_exists($key, $stats)) {
                    $result->setActualTime($stats[$key]);
                }

                $day->setWorkingTime($result);
            }
        }

        $event = new WorkingTimeYearEvent($year, $until);
        $this->eventDispatcher->dispatch($event);

        return $year;
    }

[...]

    /**
     * @param \DateTimeInterface $year
     * @param User $user
     * begin of insertion
     * filter working time overview to a specific project
     * @param int $project
     * end of insertion
     * @return array<string, int>
     */
    private function getYearStatistics(\DateTimeInterface $year, User $user, **int $project** ): array
    {
        $dateTimeFactory = DateTimeFactory::createByUser($user);
        $begin = $dateTimeFactory->createStartOfYear($year);
        $end = $dateTimeFactory->createEndOfYear($year);

        $qb = $this->timesheetRepository->createQueryBuilder('t');

        $qb
            ->select('COALESCE(SUM(t.duration), 0) as duration')
            ->addSelect('DATE(t.date) as day')
            ->where($qb->expr()->isNotNull('t.end'))
            ->andWhere($qb->expr()->between('t.begin', ':begin', ':end'))
            ->andWhere($qb->expr()->eq('t.user', ':user'))
            **/* begin of insertion
               filter working time overview to a specific project */
            ->andWhere($qb->expr()->eq('t.project', ':project'))
             /* end of insertion */**
            ->setParameter('begin', $begin)
            ->setParameter('end', $end)
            ->setParameter('user', $user->getId())
            **/* begin of insertion
              filter working time overview to a specific project */
            ->setParameter('project', $project)
            /* end of insertion */**
            ->addGroupBy('day')
        ;
        $results = $qb->getQuery()->getResult();

        $durations = [];
        foreach ($results as $row) {
            $durations[$row['day']] = (int) $row['duration'];
        }

        return $durations; // @phpstan-ignore-line
    }
}


Best regards
Klaus

**Comment by wobouser** [comment]
ps:
currently the working time recording only counts activities that have been entered.
this means that if i charge a customer 6 hours for a project, only these 6 hours are listed in the working time recording even though i was at the company for 8 hours.
conversely, this means that I am missing 2 hours

**Reply by idemus** [reply]
@wobouser 
That's why we've introduced a company internal project "Working Time" which have to be booked with working time per day in parallel. Therefore you can use the plugin "time period insert" where you can create timesheet entries in mass for recurring entries.

**Comment by wobouser** [comment]
and another small addition (but that's just optics):
the problem with the above solution is that i then have my attendance + work done listed in my weekly overview.

example:
7 hours project for the customer + 8 hours present in the company results in a total of 15 hours in my daily overview ^^

**Reply by idemus** [reply]
@wobouser
see my answer above. By introducing a project where only company internal working hours, holidays, sick leaves, etc. and no billable hours from customer projects are created, won't have that problem.

**Comment by rbuehler-teletrend** [comment]
I'm also looking for this feature to be integrated in Kimai.

**Comment by kevinpapst** [comment]
The next release will have support for that setting (see also #186 and #175).

My question to all of you who tried to add it manually: what points did you encounter that were problematic?
Would you expect that other reports (besides the working times ones) adapt to that new setting?

**Reply by idemus** [reply]
Hi Kevin,
I think all reports based on users in the menue reporting have an issue because billable hours and working hours created on that specific working hour project are sumed up. That means employees are getting results of 300 hours and more for their working hours.
So it has to be distingueshed between working hours/billable hours on custom projects and working hours on own project.

Best regards
Klaus

**Reply by kevinpapst** [reply]
Today I released the first bundle version, which has the new `System > Setting` to select a specific project, which is used for calculating the working time. The `WorkContractBundle 1.20.0` needs `Kimai 2.30.0`.

Your point is totally valid, thanks for your feedback @idemus 
This will be improved in the future. 
For now this is currently ONLY working for the work contract view: widgets and reports will still calculate all times.

**Reply by kevinpapst** [reply]
**ALL**

Please share your input here: https://github.com/kimai/Paid-plugins/issues/238

**Comment by imsgx** [comment]
We have the opposite scenario here, we have projects/activities to track future allocations.

So we need projects/activities that wont be considered in the working time calculation.

Is it possible to:
1) do not count as working time entries in specific projects or activities?
or
2) do not count as working time all entries that are after today?

thanks in advance and best regards

**Reply by kevinpapst** [reply]
1. You can configure one project that should act as working time, there no opposite "exclude list" feature.

2. Nothing after "now" should be considered working time. I just tried, its working like that.

**Reply by imsgx** [reply]
Thanks for your answer, @kevinpapst 

About 2, all these future entries since april (and also tomorrow:16/march) are being considered

![image](https://github.com/user-attachments/assets/bc2775ce-22dd-4419-9825-2561ebd3ac3f)

thanks

**Reply by imsgx** [reply]
I am not an expert, but it seems there is a OR condition in the code that checks if the month has time entries (sorry if I am totally wrong)

here: https://github.com/kimai/kimai/blob/main/templates/contract/status.html.twig
{% set showStats = now > month.month or monthIs > 0 %}

![image](https://github.com/user-attachments/assets/d031f50f-be24-4b90-85af-fd9506ef6772)

**Reply by kevinpapst** [reply]
Are those 5h entries regular timesheets?
Edit: which version of Kimai and the plugin are you using? I am not able to reproduce that behavior.
Edit 2 : please open a new bug report, this discussion is the wrong place for that topic

**Reply by imsgx** [reply]
Thanks again for your response!

I just encountered the same behavior in the trial cloud account. I can send you the URL if you'd like to check.

I also opened a new issue: https://github.com/kimai/Paid-plugins/issues/245.

Let me know if you need any further details.
